src/dynamicalsystem.jl

```diff
@@ -0,0 +1,349 @@
+using OrdinaryDiffEq, ForwardDiff, StaticArrays
+import OrdinaryDiffEq: isinplace, ODEIntegrator, step!
+import Base: eltype
+
+export dimension, state, DynamicalSystem, DS, integrator, tangent_integrator
+export ContinuousDynamicalSystem, CDS, DiscreteDynamicalSystem, DDS
+export set_parameter!, step!
+
+#######################################################################################
+#                          Basic functions and interface                              #
+#######################################################################################
+
+dimension(prob::DEProblem) = length(prob.u0)
+eltype(prob::DEProblem) = eltype(prob.u0)
```

Comment:
No... No... `DEProblem` is not the type defined in this library so you shouldn't change *any* of its API, **especially** for the Base-related APIs. (Monkey-patch is required in some cases but this is totally not the case.)  Furthermore, `eltype` mostly is for container or iterator. `DEProblem` is neither.

The second reason applies to DS.  Don't define `eltype(::DS)`.

Reply:
What? Why. 

I am using `dimension`, not `DiffEq`. What you say doesn't make sense.

Reply:
Ah, I now see why you are asking about `dimension` in discourse.  But please read carefully.  I'm talking about `Base.eltype`.  Not `dimension`.

Reply:
Yes I have already removed that, it was only for internal test.